### DISCUSSION TITLE
"Something went wrong while processing your request" after OIDC login window closes

### DISCUSSION THREAD

**Discussion by ll931217** [pocketbase/pocketbase#5352] [Q&A] [answered]
Hi!

So I am trying to implement [LINE Login's OpenID Connect](https://developers.line.biz/en/docs/line-login/integrate-line-login/#get-access-token), so far I am successful until the login window closes, I check the network tab and `/api/collections/users/auth-with-oauth2` was called, but it returned the following error:

```json
{
    "code": 400,
    "message": "Something went wrong while processing your request.",
    "data": {}
}
```

The error message doesn't have much to go by, not sure if someone can point me in the direction of where I am go to find the issue.

**Comment by ganigeorgiev** [comment]
You can try to start the executable with `--dev` and check the logs in the stderr output and/or explore the failed requests in the Admin UI > Logs section.

Note that the generic 400 error could be also returned if the OAuth2 user creation fails the Create API rule of your auth collection (eg. missing required `name` field or similar).

**Reply by ll931217** [reply]
Ah, I see the error in the logs. Thank you! Its not pocketbase related

```
failed to fetch OAuth2 user profile via https://api.line.me/oauth2/v2.1/verify (400):
{"error":"invalid_request","error_description":"Required request parameter 'access_token' for method parameter type String is not present"}
```

I presume there isn't any way to add more request parameters to the User API right? Unless going the manual route

**Reply by ganigeorgiev** [reply]
No. Are you sure that you are using the correct endpoint for the user info endpoint?

**Reply by ganigeorgiev** [reply]
I'm not sure if I'm reading correctly Line's documentation but it is possible that they don't have a user info endpoint and they allow returning the user data only as part of the `id_token` payload (which is not supported at the moment for the generic OIDC provider) - https://developers.line.biz/en/docs/line-login/verify-id-token/.

_I've added it in my todo to consider adding a toggle option for the OIDC provider to make the userinfo endpoint optional and to fetch the data from the `id_token` payload._

**Reply by ll931217** [reply]
Honestly I am not quite sure, very new to OpenID Connect. From the docs there are like 3 APIs that I can use to get user data:

- [Get profile information using ID Token](https://developers.line.biz/en/docs/line-login/verify-id-token/#get-profile-info-from-id-token) (Something I got when getting access token
- [Get user information](https://developers.line.biz/en/reference/line-login/#userinfo)
  - I tried this one but it is saying `name: Missing required value.`
- [Get user profile](https://developers.line.biz/en/reference/line-login/#get-user-profile)
  - I tried this one but it is saying `name: Missing required value.`
  
I think the second one is the most probable one, this assumption is simply based off the URL it uses `https://api.line.me/oauth2/v2.1/userinfo`

```bash
curl -v -X GET https://api.line.me/oauth2/v2.1/userinfo \
-H 'Authorization: Bearer {access token}'
````

**Reply by ganigeorgiev** [reply]
The `https://api.line.me/oauth2/v2.1/userinfo` endpoint should work.

The error "name: Missing required value." is most likely because in your auth collection the `name` field is marked "Nonempty" which will prevent creating the record when no `name` is assigned. Note that the only OAuth2 fields that are automatically assigned at the moment are:
- `email` (if returned and verified)
- `username` (if returned, doesn't exist already and satisfies with the PocketBase username format, otherwise generates a random one)

With the next release there is option to manually map from the Admin UI the standardized OAuth2 user fields to their corresponding record equivalents but for now you'll have to do it manually either via [OnRecordBeforeAuthWithOAuth2Request hook](https://pocketbase.io/docs/go-event-hooks/#onrecordbeforeauthwithoauth2request) (_this could be tricky though with the current hooks execution since you'll have to initialize a blank empty record and fill the fields in a similar manner to the [form](https://github.com/pocketbase/pocketbase/blob/498a21b020d0f171ba7aaf4b91f46c76b2460635/forms/record_oauth2_login.go#L205)_) or client-side with and extra update request (_this is easier but you'll have to remove the non-empty constraint_).

**Reply by ll931217** [reply]
By "automatically assigned", do you mean what is returned by the `userinfo` endpoint? So if `email` and/or `username` is returned then it will automatically add it to PB and the rest of the fields won't be mapped?

**Reply by ganigeorgiev** [reply, marked as answer]
> ... it will automatically add it to PB and the rest of the fields won't be mapped?

Yes. This is because the `name`, `avatar`, etc. fields are custom and user defined (for example you can delete the default `name` field and instead have `firstName` and `lastName` `text` fields).

Again with the next release this is simplified in both in terms of handling it on the server-side (the OAuth2 user creation will fire also the regular Record create request hook) and in the Admin UI you also will have an option to manually specify the extra mapping (at least for the normalized `auth.AuthUser` fields). In v0.23 it will look like this:

![screenshot](https://i.imgur.com/rXCHIgD.png)

For now the easiest option is to disable the Nonempty name constraint and send an update request after the OAuth2 auth (keep in mind that not all OAuth2 providers return or allow retrieving the name of the user).

```js
const authData = await pb.collection("users").authWithOAuth2({ ... })

// the OAuth2 user data is returned under the authData.meta prop
await pb.collection("users").update(authData.record.id, { name: authData.meta.something })
```

**Reply by ll931217** [reply]
Awesome! Is there like an estimated date v0.23 will be released?

**Reply by ganigeorgiev** [reply]
I expect to finalize it within the next 1 or 2 months at worst, but there are no fixed ETAs.

**Reply by ll931217** [reply]
Okay thanks so much!

**Reply by ll931217** [reply]
Hi @ganigeorgiev do you know if it's possible to prevent OIDC from opening a pop up to login? Since LINE uses an internal browser that blocks pop ups

**Reply by ganigeorgiev** [reply]
I don't exactly understand what do you mean by _"LINE uses an internal browser that blocks pop ups"_, but if you don't want to open the provider's login page in a new window or tab then you'll have to use the "Manual code exchange" flow (the second tab in https://pocketbase.io/docs/authentication/#oauth2-integration).

If you have further questions, please open a new Q&A discussion instead.
